eval/topdown_test.go

```diff
@@ -0,0 +1,540 @@
+	tests := []struct {
+		note     string
+		rule     string
+		expected string
+	}{
+		{"undefined", "p = null :- false", ""}, // "" will be converted to Undefined
+		{"null", "p = null :- true", "null"},
+		{"bool: true", "p = true :- true", "true"},
+		{"bool: false", "p = false :- true", "false"},
+		{"number: 3", "p = 3 :- true", "3"},
+		{"number: 3.0", "p = 3.0 :- true", "3.0"},
+		{"number: 66.66667", "p = 66.66667 :- true", "66.66667"},
+		{`string: "hello"`, `p = "hello" :- true`, `"hello"`},
+		{`string: ""`, `p = "" :- true`, `""`},
+		{"array: [1,2,3,4]", "p = [1,2,3,4] :- true", "[1,2,3,4]"},
+		{"array: []", "p = [] :- true", "[]"},
+		{`object/nested composites: {"a": [1], "b": [2], "c": [3]}`,
+			`p = {"a": [1], "b": [2], "c": [3]} :- true`,
+			`{"a": [1], "b": [2], "c": [3]}`},
+	}
+
+	data := loadSmallTestData()
+
+	for i, tc := range tests {
+
+		ctx := &TopDownContext{
+			Rule:     parseRule(tc.rule),
+			Store:    NewStorageFromJSONObject(data),
+			Bindings: make(map[opalog.Var]opalog.Value),
+		}
+
+		expected := loadExpectedResult(tc.expected)
+		result, err := TopDownQuery(ctx)
+
+		if err != nil {
+			t.Errorf("Test case %d (%v): unexpected error: %v", i+1, tc.note, err)
+			continue
+		}
+
+		if !reflect.DeepEqual(result, expected) {
+			t.Errorf("Test case %d (%v): expected %v but got: %v", i+1, tc.note, expected, result)
+		}
+	}
+
+}
+
+func TestTopDownSetDoc(t *testing.T) {
+
+	tests := []struct {
+		note     string
+		rule     string
+		expected string
+	}{
+		{"array values", "p[x] :- a[i] = x", `[1, 2, 3, 4]`},
+		{"array indices", "p[x] :- a[x] = _", `[0, 1, 2, 3]`},
+		{"object keys", "p[x] :- b[x] = _", `["v1", "v2"]`},
+		{"object values", "p[x] :- b[i] = x", `["hello", "goodbye"]`},
+		{"nested composites", "p[x] :- f[i] = x", `[{"xs": [1.0], "ys": [2.0]}, {"xs": [2.0], "ys": [3.0]}]`},
+		{"deep ref/heterogeneous", "p[x] :- c[i][j][k] = x", `[null, 3.14159, true, false, true, false, "foo"]`},
+	}
+
+	data := loadSmallTestData()
+
+	for i, tc := range tests {
+		ctx := &TopDownContext{
+			Rule:     parseRule(tc.rule),
+			Store:    NewStorageFromJSONObject(data),
+			Bindings: make(map[opalog.Var]opalog.Value),
+		}
+
+		expected := loadExpectedResult(tc.expected)
+		result, err := TopDownQuery(ctx)
+
+		if err != nil {
+			t.Errorf("Test case %d (%v): unexpected error: %v", i+1, tc.note, err)
+			continue
+		}
+
+		sort.Sort(ResultSet(result.([]interface{})))
+
+		if !reflect.DeepEqual(result, expected) {
+			t.Errorf("Test case %d (%v): expected %v but got: %v", i+1, tc.note, expected, result)
+		}
+	}
+
+}
+
+func TestTopDownObjectDoc(t *testing.T) {
+	tests := []struct {
+		note     string
+		rule     string
+		expected interface{}
+	}{
+		{"identity", "p[k] = v :- b[k] = v", `{"v1": "hello", "v2": "goodbye"}`},
+		{"composites", "p[k] = v :- d[k] = v", `{"e": ["bar", "baz"]}`},
+		{"non-string key", "p[k] = v :- a[k] = v", fmt.Errorf("cannot produce object with non-string key: 0")},
+	}
```

Comment:
Make sense to add tests with variables in the body that do NOT occur in the head?  Maybe a join.  Might need different data.
p[k] = v :- a[k] = v2, b[v2] = v